### DISCUSSION TITLE
Special characters in sqlserver connector

### DISCUSSION THREAD

**Discussion by erekkiiker** [slingdata-io/sling-cli#331] [Q&A] [answered]
I have a source sqlserver that has tables with spaces in the name 
When I run sling conns discover for such table (wrapped in "schema.my table") it finds the object.
However when I do sling run (wrapping the name same way in quotes) I get error.
11:49PM INF execution failed

fatal:
~ SQL Error
mssql: Incorrect syntax near 'Charges'.

Is there a way to replicate objects with such names or is this some known limitation?

**Comment by flarco** [comment]
Could you share your config?

**Reply by erekkiiker** [reply]
Discover that works I run using
sling conns discover onprem_sql -p "DM.UK_Export_Auto Misc Charges Vendor Atlas_v6"
And run that fails I tried with 
sling run --src-conn onprem_sql --src-stream "DM.UK_Export_Auto Misc Charges Vendor Atlas_v6" --tgt-conn azure_sql --tgt-object egdp.UK_Export_Auto_Misc_Charges_Vendor_Atlas_v6 --mode full-refresh
Tried also without quotes for src-stream but same result.
For the tables without any spaces in names the connections work.

**Reply by flarco** [reply]
Can you try with single quote and double quotes? Like this:

`sling run --src-conn onprem_sql --src-stream '"DM"."UK_Export_Auto Misc Charges Vendor Atlas_v6"' --tgt-conn azure_sql --tgt-object '"egdp"."UK_Export_Auto_Misc_Charges_Vendor_Atlas_v6"' --mode full-refresh`

**Reply by erekkiiker** [reply]
Yes, this worked. Thank you.
How would I describe this table name in replication yaml? Same syntax doesn't seem to work
streams:
  '"DM"."UK_Export_Auto Misc Charges Vendor Atlas_v6"':
    object: egdp.{stream_table}
    mode: full-refresh

**Comment by flarco** [comment]
can you share your replication yaml?

**Reply by erekkiiker** [reply]
I've tried to set it up like this (and different variations with single and double quotes)

source: onprem_sql
target: azure_sql

defaults:
  object: '{target_schema}.{stream_table}'
  mode: full-refresh
  target_options:
    file_max_rows: 100000 # will split files into folder
    file_max_bytes: 32768000 

streams:
  '"DM"."UK_Export_Auto Misc Charges Vendor Atlas_v6"':
    object: 'egdp.{stream_table}'

**Reply by flarco** [reply]
can you share the log output with debug mode (run with flag`-d`)?

**Reply by erekkiiker** [reply]
Here's the debug output

PS C:\Users\x\source\repos\sling> sling run -d -r .\egdp_replicate.yaml
2024-06-30 20:30:41 INF Sling Replication [1 streams] | on_prem_sql -> azure_sql

2024-06-30 20:30:41 INF [1 / 1] running stream "DM"."UK_Export_Auto Misc Charges Vendor Atlas_v6"
2024-06-30 20:30:41 DBG Sling version: 1.2.11 (windows amd64)
2024-06-30 20:30:41 DBG type is db-db
2024-06-30 20:30:41 DBG using source options: {"empty_as_null":false,"null_if":"NULL","datetime_format":"AUTO","max_decimals":-1}
2024-06-30 20:30:41 DBG using target options: {"datetime_format":"auto","file_max_rows":100000,"file_max_bytes":32768000,"max_decimals":-1,"use_bulk":true,"add_new_columns":true,"adjust_column_type":false,"column_casing":"source"}
2024-06-30 20:30:41 INF connecting to source database (sqlserver)
2024-06-30 20:30:41 DBG opened "sqlserver" connection (conn-sqlserver-NF6)
2024-06-30 20:30:41 INF connecting to target database (sqlserver)
2024-06-30 20:30:41 DBG opened "sqlserver" connection (conn-sqlserver-Z68)
2024-06-30 20:30:42 INF reading from source database
2024-06-30 20:30:42 INF execution failed
2024-06-30 20:30:42 INF ~ could not parse target table name
invalid table name

2024-06-30 20:30:42 INF Sling Replication Completed in 1s | on_prem_sql -> azure_sql | 0 Successes | 1 Failures

fatal:
--- proc.go:267 main ---
--- sling_cli.go:442 main ---
--- sling_cli.go:474 cliInit ---
--- cli.go:286 CliProcess ---
~ failure running replication (see docs @ https://docs.slingdata.io/sling-cli)
--- sling_run.go:188 processRun ---

--------------------------- "DM"."UK_Export_Auto Misc Charges Vendor Atlas_v6" ---------------------------
--- task_run.go:99 func1 ---
~ execution failed
--- task_run.go:523 runDbToDb ---
~ Could not ReadFromDB
--- task_run_read.go:50 ReadFromDB ---
~ could not get format map for sql
--- config.go:709 GetFormatMap ---
~ could not parse target table name
--- schemata.go:559 ParseTableName ---
invalid table name

**Reply by flarco** [reply]
I think this is a YAML space indentation issue. Can you copy/paste this exactly:

```yaml
source: onprem_sql
target: azure_sql

defaults:
  object: '{target_schema}.{stream_table}'
  mode: full-refresh
  target_options:
    file_max_rows: 100000 # will split files into folder
    file_max_bytes: 32768000

streams:
  '"DM"."UK_Export_Auto Misc Charges Vendor Atlas_v6"':
    object: 'egdp.{stream_table}'
```

**Reply by flarco** [reply]
You can also format it in JSON if you're more comfortable:

```json
{
  "source": "onprem_sql", 
  "target": "azure_sql", 
  "defaults": {
    "target_options": {
      "file_max_bytes": 32768000, 
      "file_max_rows": 100000
    }, 
    "object": "{target_schema}.{stream_table}", 
    "mode": "full-refresh"
  }, 
  "streams": {
    "\"DM\".\"UK_Export_Auto Misc Charges Vendor Atlas_v6\"": {
      "object": "egdp.{stream_table}"
    }
  }
}
```

**Reply by erekkiiker** [reply]
I copied yaml from you message and I got the same result. And when I compared to my original one it was identical, including all the spaces. 
I notices now that when I copied yaml content it lost the indentation. 
Replication with yaml works with other tables that don't have any spaces in their name

**Reply by erekkiiker** [reply]
Tried also json but still the same result

2024-06-30 22:45:06 INF reading from source database
2024-06-30 22:45:06 INF execution failed

2024-06-30 22:45:07 INF ~ could not parse target table name
invalid table name
fatal:
--- proc.go:267 main ---
--- sling_cli.go:442 main ---
--- sling_cli.go:474 cliInit ---
--- cli.go:286 CliProcess ---
~ failure running replication (see docs @ https://docs.slingdata.io/sling-cli)
--- sling_run.go:188 processRun ---

--------------------------- "DM"."UK_Export_Auto Misc Charges Vendor Atlas_v6" ---------------------------
--- task_run.go:99 func1 ---
~ execution failed
--- task_run.go:523 runDbToDb ---
~ Could not ReadFromDB
--- task_run_read.go:50 ReadFromDB ---
~ could not get format map for sql
--- config.go:709 GetFormatMap ---
~ could not parse target table name
--- schemata.go:559 ParseTableName ---
invalid table name

**Reply by flarco** [reply]
Thanks, there is a bug, will fix.

**Reply by flarco** [reply, marked as answer]
It's because of the spaces in the stream name, `egdp.{stream_table}` gets rendered to `egdp.uk_export_auto misc charges vendor atlas_v6` which silently errored since it is not a valid table name.
 
Fixed here: https://github.com/slingdata-io/sling-cli/pull/332/commits/ff7d4d97bbe3cf2bb3bcfd33f1d51382ada75683
Feel free to build the binary from that branch.
For now, you can simply specify the object name for that stream without using  `{stream_table}` (`"egdp"."uk_export_auto_misc_charges_vendor_atlas_v6"`).

**Reply by erekkiiker** [reply]
Thanks.
